lightning/src/ln/channelmanager.rs

```diff
@@ -2059,10 +2128,57 @@
 		} else if has_err {
 			Err(PaymentSendFailure::AllFailedRetrySafe(results.drain(..).map(|r| r.unwrap_err()).collect()))
 		} else {
-			Ok(())
+			Ok(payment_id)
 		}
 	}
 
+	/// Retries a payment along the given [`Route`].
+	///
+	/// Errors returned are a superset of those returned from [`send_payment`], so see
+	/// [`send_payment`] documentation for more details on errors. This method will also error if the
+	/// retry amount puts the payment nontrivially over the payment's total amount, or if the payment
+	/// for the given `payment_id` cannot be found (likely due to timeout or success).
+	///
+	/// [`send_payment`]: [`ChannelManager::send_payment`]
+	pub fn retry_payment(&self, route: &Route, payment_id: PaymentId) -> Result<(), PaymentSendFailure> {
+		for path in route.paths.iter() {
+			if path.len() == 0 {
+				return Err(PaymentSendFailure::ParameterError(APIError::APIMisuseError {
+					err: "length-0 path in route".to_string()
+				}))
+			}
+		}
+
+		let (total_msat, payment_hash, payment_secret) = {
+			let outbounds = self.pending_outbound_payments.lock().unwrap();
+			if let Some(payment) = outbounds.get(&payment_id) {
+				match payment {
+					PendingOutboundPayment::Retryable {
+						total_msat, payment_hash, payment_secret, pending_amt_msat, ..
+					} => {
+						let retry_amt_msat: u64 = route.paths.iter().map(|path| path.last().unwrap().fee_msat).sum();
+						if retry_amt_msat + *pending_amt_msat > *total_msat * 11 / 10 {
```

Comment:
The excess threshold (10%) should be parametrized in a constant

Reply:
Given this function will mainly be used by Jeff's retryer, so he'd have to add this parameter as well for it to be useful, I'm in favor of punting this to follow up depending on whether we get complaints :stuck_out_tongue: